### ISSUE TITLE
Profiling disabled due to missing extension

### ISSUE THREAD

**Issue by theocodes** [DataDog/dd-trace-rb#2067]
Hi 👋 

Came across the following while attempting to set up profilingf or a rack app

```
W, [2022-06-06T10:16:18.000528 #130]  WARN -- ddtrace: [ddtrace] Profiling was requested but is not supported, profiling disabled: There was an error loading the profiling native extension due to 'RuntimeError Failure to load ddtrace_profiling_native_extension.2.7.6_x86_64-linux due to libddprof_ffi.so: cannot open shared object file: No such file or directory' at '/app/vendor/bundle/ruby/2.7.0/gems/ddtrace-1.1.0/lib/datadog/profiling/load_native_extension.rb:22:in `<top (required)>''
```

After some digging I can see this (at least the error message) may be caused by the 1.1.0 release which added `libddprof` as a dependency. However, it's not clear to me whether this is intended or a bug..

[The PR](https://github.com/DataDog/dd-trace-rb/pull/2028) Seems to suggest x86_64 linux is supported and that is exactly what I am using and yet I get the error.

Would youf folks be able to shed some light on it?

Thanks!

**Comment by ivoanjo** [comment]
Hey @theocodes thanks for the report and interest in the Profiler!

This is definitely not expected, as indeed x86_64 Linux is fully supported. Ever since the PR you mentioned (#2028) the profiler gets compiled and linked to the libddprof library, which is shipped in the [`libddprof`](https://rubygems.org/gems/libddprof) gem.

The weird part is that the error you shared seems to indicate that the profiler did get successfully compiled and linked during installation (otherwise the error message would be different), so whatever went wrong, it went wrong after `ddtrace` was installed.

In particular, it looks like the `libddprof` gem was present at installation time, but seemed to not be found anymore at execution time, hence the `libddprof_ffi.so: cannot open shared object file: No such file or directory` that you reported.

* :one: Could you share the output of running the following command:
```ldd `gem which ddtrace_profiling_native_extension.2.7.6_x86_64-linux.so` ```?

  This should help us figure out where the `libddprof_ffi.so` was expected to be found in your installation.

* :two: Could you share a few more details about how your app and its dependencies gets installed and deployed? Hopefully I can reproduce the issue on my side as well.

**Comment by theocodes** [comment]
Hi @ivoanjo 

Thanks for looking into this!

I'm not sure what went wrong there but I'm unable to reproduce it now - everything seems to be working.. 🤔 

Maybe let's close this issue for now and I'll let you folks know if it happens again?

**Comment by ivoanjo** [comment]
Great! Feel free to reach out if there's anything we can do to help, and happy profiling :)

**Comment by Baron-burton** [comment]
Hey @ivoanjo, I've hit this error myself today. 

I get the following output from the command you suggested to run:

```	
linux-vdso.so.1 (0x00007fff69f3e000)
libruby.so.3.1 => not found
libddprof_ffi.so => not found
libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f1ff6939000)
libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f1ff6747000)
/lib64/ld-linux-x86-64.so.2 (0x00007f1ff6a97000)
```

Currently running Ruby 3.1.2

**Comment by Baron-burton** [comment]
I've tested reverting ddtrace back to 1.0.0 and this gets rid of the errors. 

I mentioned that we're running Ruby 3.1.2 but we're also running Rails 7.0.3

**Comment by ivoanjo** [comment]
Thanks @Baron-burton for the extra info. Yes, this issue was the result of a new change in 1.1, but obviously not intended to break ;)

The output you shared is quite interesting/helpful, since it does confirm that whatever paths were observed during installation to reach `libruby.so.3.1` and `libddprof_ffi.so` don't seem to be there anymore.

So I can investigate further, could you provide the output of the following:

* :one:  ```ldd `which ruby` ```
  Usually Ruby is installed so that the `ruby` command is just a tiny executable that is linked to `libruby.so`, so this should print where it can be found
 
* :two:  `gem contents libddprof`
  This will show where the `libddprof` gem (a dependency of ddtrace) files are installed in your system

* :three: ```strings `gem which ddtrace_profiling_native_extension.3.1.2_x86_64-linux.so` | grep -e "lib\(ruby\|ddprof\)"```
  This command may need to be tweaked to your exact Ruby version; it will print the paths that were used when installing the profiling native extension.
  
And a couple of questions:

* :four: Do you happen to be using a public Ruby build or docker image that you could share? That would help a lot in reproducing this issue

* :five: If the above isn't possible, could you share some details on how you're installing your dependencies and deploying your application?

Thanks again, and I'll get to the bottom of this issue soon :)

**Comment by Baron-burton** [comment]
Hey @ivoanjo, thanks for getting back to me

1️⃣ ldd `which ruby`

```
linux-vdso.so.1 (0x00007fff9ebfe000)
libruby.so.3.1 => not found
libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007fb4d524d000)
libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fb4d522a000)
librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007fb4d5220000)
libgmp.so.10 => /lib/x86_64-linux-gnu/libgmp.so.10 (0x00007fb4d519c000)
libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fb4d5194000)
libcrypt.so.1 => /lib/x86_64-linux-gnu/libcrypt.so.1 (0x00007fb4d5159000)
libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fb4d500a000)
libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fb4d4e18000)
/lib64/ld-linux-x86-64.so.2 (0x00007fb4d5276000)
```

2️⃣ gem contents libddprof

```
/app/vendor/bundle/ruby/3.1.0/gems/libddprof-0.6.0.1.0-x86_64-linux/lib/libddprof.rb
/app/vendor/bundle/ruby/3.1.0/gems/libddprof-0.6.0.1.0-x86_64-linux/lib/libddprof/version.rb
/app/vendor/bundle/ruby/3.1.0/gems/libddprof-0.6.0.1.0-x86_64-linux/vendor/libddprof-0.6.0/x86_64-linux-musl/libddprof-x86_64-alpine-linux-musl/LICENSE
/app/vendor/bundle/ruby/3.1.0/gems/libddprof-0.6.0.1.0-x86_64-linux/vendor/libddprof-0.6.0/x86_64-linux-musl/libddprof-x86_64-alpine-linux-musl/LICENSE-3rdparty.yml
/app/vendor/bundle/ruby/3.1.0/gems/libddprof-0.6.0.1.0-x86_64-linux/vendor/libddprof-0.6.0/x86_64-linux-musl/libddprof-x86_64-alpine-linux-musl/NOTICE
/app/vendor/bundle/ruby/3.1.0/gems/libddprof-0.6.0.1.0-x86_64-linux/vendor/libddprof-0.6.0/x86_64-linux-musl/libddprof-x86_64-alpine-linux-musl/include/ddprof/ffi.h
/app/vendor/bundle/ruby/3.1.0/gems/libddprof-0.6.0.1.0-x86_64-linux/vendor/libddprof-0.6.0/x86_64-linux-musl/libddprof-x86_64-alpine-linux-musl/lib/libddprof_ffi.so
/app/vendor/bundle/ruby/3.1.0/gems/libddprof-0.6.0.1.0-x86_64-linux/vendor/libddprof-0.6.0/x86_64-linux-musl/libddprof-x86_64-alpine-linux-musl/lib/pkgconfig/ddprof_ffi_with_rpath.pc
/app/vendor/bundle/ruby/3.1.0/gems/libddprof-0.6.0.1.0-x86_64-linux/vendor/libddprof-0.6.0/x86_64-linux/libddprof-x86_64-unknown-linux-gnu/LICENSE
/app/vendor/bundle/ruby/3.1.0/gems/libddprof-0.6.0.1.0-x86_64-linux/vendor/libddprof-0.6.0/x86_64-linux/libddprof-x86_64-unknown-linux-gnu/LICENSE-3rdparty.yml
/app/vendor/bundle/ruby/3.1.0/gems/libddprof-0.6.0.1.0-x86_64-linux/vendor/libddprof-0.6.0/x86_64-linux/libddprof-x86_64-unknown-linux-gnu/NOTICE
/app/vendor/bundle/ruby/3.1.0/gems/libddprof-0.6.0.1.0-x86_64-linux/vendor/libddprof-0.6.0/x86_64-linux/libddprof-x86_64-unknown-linux-gnu/include/ddprof/ffi.h
/app/vendor/bundle/ruby/3.1.0/gems/libddprof-0.6.0.1.0-x86_64-linux/vendor/libddprof-0.6.0/x86_64-linux/libddprof-x86_64-unknown-linux-gnu/lib/libddprof_ffi.so
/app/vendor/bundle/ruby/3.1.0/gems/libddprof-0.6.0.1.0-x86_64-linux/vendor/libddprof-0.6.0/x86_64-linux/libddprof-x86_64-unknown-linux-gnu/lib/pkgconfig/ddprof_ffi_with_rpath.pc
```

3️⃣ strings `gem which ddtrace_profiling_native_extension.3.1.2_x86_64-linux.so` | grep -e "lib\(ruby\|ddprof\)"

```
libruby.so.3.1
libddprof_ffi.so
/tmp/build_9b486d62/vendor/ruby-3.1.2/lib:/tmp/build_9b486d62/vendor/bundle/ruby/3.1.0/gems/libddprof-0.6.0.1.0-x86_64-linux/vendor/libddprof-0.6.0/x86_64-linux/libddprof-x86_64-unknown-linux-gnu/lib/pkgconfig/../../lib:${ORIGIN}/../lib
```

4️⃣ Ruby installation

We rely on Heroku to install our Ruby version. This is done by simply specifying the version in our Gemfile, Heroku picks it up from there and installs the required version.

**Comment by seuros** [comment]
Do we have any timeline when 1.2.0 will be released ? 

Btw : This issue happens in AWS Elastic Beanstalks too.

**Comment by ivoanjo** [comment]
Thanks for the patience! We expect to have it out within the next two weeks. Fixing this and getting 1.2.0 is at the top of the priority list.

> Btw : This issue happens in AWS Elastic Beanstalks too.

Interesting, I did not know that! I've been able to reproduce the issue on Heroku, and I'll additionally validate that the fix also works on Elastic Beanstalk.

**Comment by ivoanjo** [comment]
Thanks folks for the patience so far! I've merged in a fix for this, and the plan is to release it as part of ddtrace 1.2.0 next week.

**Comment by ivoanjo** [comment]
👋 @Baron-burton @seuros the fix for this issue has been released in [v1.2.0](https://github.com/DataDog/dd-trace-rb/releases/tag/v1.2.0).
Thanks for the help with the investigation! 🙇

**Comment by ivoanjo** [comment]
@seuros damn, big apologies for disappointing you twice! I've been reviewing the code and have a few suspicions; I've opened https://github.com/DataDog/dd-trace-rb/issues/2151 to track this issue.
